finality-aleph/src/sync/data.rs

```diff
@@ -140,17 +173,19 @@ impl<J: Justification> Encode for VersionedNetworkData<J> {
         match self {
             Other(version, payload) => encode_with_version(*version, payload),
             V1(data) => encode_with_version(Version(1), &data.encode()),
+            V2(data) => encode_with_version(Version(2), &data.encode()),
         }
     }
 }
 
-impl<J: Justification> Decode for VersionedNetworkData<J> {
+impl<B: Block, J: Justification> Decode for VersionedNetworkData<B, J> {
     fn decode<I: CodecInput>(input: &mut I) -> Result<Self, CodecError> {
         use VersionedNetworkData::*;
         let version = Version::decode(input)?;
         let num_bytes = ByteCount::decode(input)?;
         match version {
-            Version(1) => Ok(V1(NetworkData::decode(input)?)),
+            Version(1) => Ok(V1(NetworkDataV1::decode(input)?)),
+            Version(2) => Ok(V2(NetworkData::decode(input)?)),
```

Comment:
Suggestion: Are we planning to have more versions here? Or is this only for the purpose of the `rolling-upgrade` procedure and we only need last two versions? If so, then you probably could simplify all those constraints, starting with this `VersionedNetworkData` (generic on any two params V1, V2  instead of that B, J) and `VersionWrapper` (no dependency on particular `NetworkData`). I can see why these constraints are becoming annoying :)

Reply:
Honestly? No idea. We'll see in a few months :)